Allow specifying the response classifier in RetryPolicy (#37715)
Azure/azure-sdk-for-net: sdk/core/Azure.Core/samples/Configuration.md

};
```

> Another customization that is possible is to easily configure what exceptions or status codes 

Reviewer: Possibly silly question, but if the primary motivating scenario for this is extending/overriding the set of retriable status codes, would a simpler API be to provide a list of those status codes?

Reviewer: I see, so you were interested in not just status codes, but also exceptions that indicated the need to retry?  Or were you interested in overriding IsErrorResponse here as well?  If the latter is true, I'm not sure the current approach will allow that because it's not used elsewhere in the pipeline.

Actually, thinking about this more, @jsquire, is there a reason you need the classifier to be localized to the retry policy? Or could you set one another way, like through the pipeline or on the message or request context?

Is this the first time we're making it possible for an end-user customer and not a library author to set response classifier?  Or, the first time outside of request context?  If so, we may want to consider those implications as well.

Reviewer: Also, I'm wondering - given the complexity we're discovering, if this is needed internally for testing and not by customers, would it make sense to address it differently for this case?

Author: I think this is a clear gap in the functionality of RetryPolicy. A very common scenario is that users want to configure what will be retried. ResponseClassifier provides a super easy interface to accomplish this, and it is a building block of RetryPolicy.

> Another customization that is possible is to easily configure what exceptions or status codes 
> are retried. By default, this is configured in the `ResponseClassifier` type, but it can be 

Reviewer 2: ```suggestion
If you only need configure which exceptions or status codes are retried, this is also supported.
By default, this is configured in the `ResponseClassifier` type, but it can be 
```

overridden by setting the `ResponseClassifier` property on the `RetryPolicyOptions` when 
constructing a `RetryPolicy`. In the example below, we create a custom response classifier that 
we use in our custom `RetryPolicy`.
```C# Snippet:CustomResponseClassifier
public class CustomResponseClassifier : ResponseClassifier
{
    public override bool IsRetriableResponse(HttpMessage message)
    {
        switch (message.Response.Status)
        {
            case 404:
                return true;
            default:

>                 return base.IsRetriableResponse(message);

Reviewer: Question: the note on line 204 re: when library-specific response classifiers will and won't be respected makes me wonder if we should consider shifting the design so that it always calls through to the base if the caller doesn't handle every case.  This would make it less error prone if someone forgot to call the base, but it may not be worth the complexity and/or may not be the desired behavior.

Reviewer 3: This was based on my port of test infrastructure in #37568.  I wanted to include not only additional response codes but also custom error logic - basically, I wanted the full control of a classifier and ability to control the logic without having to write a custom retry policy just to inject one.

Author: Yeah, I don't think we do this in general as it relies on a more complex surface. It is the overrider's responsibility to handle the different statuses as they see fit.
